cloudant-sync-datastore-android/src/main/java/com/cloudant/sync/replication/PeriodicReplicationService.java

```diff
@@ -0,0 +1,329 @@
+package com.cloudant.sync.replication;
+
+import android.app.AlarmManager;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.os.Handler;
+import android.os.IBinder;
+import android.os.Looper;
+import android.os.Message;
+import android.os.SystemClock;
+import android.util.Log;
+
+/**
+ * This {@link android.app.Service} is an abstract class that is the basis for creating a service
+ * that performs periodic replications (i.e. replications that occur at regular intervals). The
+ * period between replications may be varied depending on whether other application components
+ * are bound to the service or not, so as to allow for more frequent replications when an app
+ * is in active use and less frequent replications the rest of the time.
+ *
+ * @param <T> The {@link PeriodicReplicationReceiver} associated with this Service that is
+ *           responsible for handling the alarms triggered by the {@link AlarmManager} at
+ *           the intervals when replication is required and handles resetting of alarms after
+ *           reboot of the device.
+ */
+public abstract class PeriodicReplicationService<T extends PeriodicReplicationReceiver>
+    extends ReplicationService {
+
+    /* Name of the SharedPreferences file used to store alarm times. We store the alarm
+     * times in preferences so we can reset the alarms as accurately as possible after reboot
+     * and so we can adjust alarm times when components bind to or unbind from this Service. */
+    private static final String PREFERENCES_FILE_NAME = "com.cloudant.preferences";
```

Comment:
Good idea. My preference would be to add the fully qualified package name to the field names in the file rather than changing the filename. My personal feeling is that developers are more likely to be accepting of a single Cloudant specific preferences file than multiple files, so having a generic name like `com.cloudant.preferences` allows us to use it if we ever need to for anything else, but by looking at the fields inside you get the traceability to where the preference is used.
Done in c7b0975.